cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -0,0 +1,299 @@
+  switch (tensor->type_id()) {
+    case Type::UINT8: {
+      value_array = std::make_shared<UInt8Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT8: {
+      value_array = std::make_shared<Int8Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT16: {
+      value_array = std::make_shared<UInt16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT16: {
+      value_array = std::make_shared<Int16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT32: {
+      value_array = std::make_shared<UInt32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT32: {
+      value_array = std::make_shared<Int32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::HALF_FLOAT: {
+      value_array = std::make_shared<HalfFloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::FLOAT: {
+      value_array = std::make_shared<FloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::DOUBLE: {
+      value_array = std::make_shared<DoubleArray>(tensor->size(), tensor->data());
+      break;
+    }
+    default: {
+      return Status::NotImplemented("Unsupported tensor type: ",
+                                    tensor->type()->ToString());
+    }
+  }
+  arr = std::make_shared<FixedSizeListArray>(ext_type->storage_type(), tensor->shape()[0],
+                                             value_array);
+  auto ext_data = arr->data();
+  ext_data->type = ext_type;
+  return MakeArray(ext_data);
+}
+
+Result<std::shared_ptr<Tensor>> FixedShapeTensorType::ToTensor(
+    std::shared_ptr<Array> arr) {
+  // To convert an array of n dimensional tensors to a n+1 dimensional tensor we
+  // interpret the array's length as the first dimension the new tensor. Further, we
+  // define n+1 dimensional tensor's strides by front appending a new stride to the n
+  // dimensional tensor's strides.
+
+  ARROW_CHECK(is_tensor_supported(this->value_type_->id()));
+
+  ARROW_DCHECK_EQ(arr->null_count(), 0) << "Null values not supported in tensors.";
+  auto ext_arr = internal::checked_pointer_cast<FixedSizeListArray>(
+      internal::checked_pointer_cast<ExtensionArray>(arr)->storage());
+
+  std::vector<int64_t> shape = this->shape();
+  shape.insert(shape.begin(), 1, arr->length());
+
+  std::vector<int64_t> tensor_strides = this->strides();
+  tensor_strides.insert(tensor_strides.begin(), 1, arr->length() * tensor_strides[0]);
+
+  std::shared_ptr<Buffer> buffer = ext_arr->values()->data()->buffers[1];
+  return *Tensor::Make(ext_arr->value_type(), buffer, shape, tensor_strides, dim_names());
```

Comment:
It might be best to let `Tensor::Make` recompute the strides here (based on the `shape`). We would get a multiplication overflow check baked in. You'd just have to pass an empty vector.

Also, would `permutation` play a role in this conversion? i.e. would `shape` potentially need to permuted before passing it to the `Tensor` to reflect a row-major layout in the buffer`? (actually not sure)

Reply:
We chose to represent strides with a permutation of shape. Since we're starting from an array here that has an arbitrary strides configuration we'd want to use ones computed from the permutation (as is) and not ones computed from shape (which would be relying on `Tensor::Make`).
I've added a stride overflow check and put `Tensor::Make` into `ARROW_ASSIGN_OR_RAISE` to propagate errors.

Reply:
cc @jorisvandenbossche

Reply:
> We chose to represent strides with a permutation of shape. Since we're starting from an array here that has an arbitrary strides configuration we'd want to use ones computed from the permutation (as is) and not ones computed from shape (which would be relying on `Tensor::Make`).

Won't that give wrong results? (I am not super familiar with the Tensor implementation , though) 
The actual buffer that is being passed to Tensor::Make is row-major, and the doc comment of Tensor::Make says about the `strides` param that "if this is empty, the data assumed to be row-major". So if it is not empty and not a standard strides (but actually permuted strides), won't that give a wrong interpretation of the data?

(also, if you edit strides, I would assume you also have to change the shape? And the dim_names)

Is there a test that covers this? I only see one ToTensor call in the tests (that is not checking an error) in `RoundtripTensor`, and I think that is not using custom strides?

(personally, I would just leave any handling of strides/permutation to the user/application, and they can do that after getting the Tensor / numpy array)

Reply:
> Won't that give wrong results? (I am not super familiar with the Tensor implementation , though) The actual buffer that is being passed to Tensor::Make is row-major, and the doc comment of Tensor::Make says about the `strides` param that "if this is empty, the data assumed to be row-major". So if it is not empty and not a standard strides (but actually permuted strides), won't that give a wrong interpretation of the data?

Tensor's strides are arbitrary as long they are valid ([CheckTensorStridesValidity](https://github.com/apache/arrow/blob/2ff4e3a2523bd0c58168d6ca4bcb14f45393ff2b/cpp/src/arrow/tensor.cc#L154-L191)). If we have elements with non-trivial strides we'd probably want to keep those (adding the 0-th stride to account for the "array length dimension") when calling `ToTensor`? If we want to avoid this we could also permute dim names so as to map to the physical layout as `Tensor` doesn't have permutation property (should we consider adding it?).

> (also, if you edit strides, I would assume you also have to change the shape? And the dim_names)

`ToTensor` currently amends `strides`, `shape` and `dim_names`.

> Is there a test that covers this? I only see one ToTensor call in the tests (that is not checking an error) in `RoundtripTensor`, and I think that is not using custom strides?

I'll add some. They will more or less just trigger error checks in `Tensor::Make` .

> (personally, I would just leave any handling of strides/permutation to the user/application, and they can do that after getting the Tensor / numpy array)

I think without handling strides in `ToTensor` we can't roundtrip numpy arrays with non-trivial arrays?

Reply:
> > (also, if you edit strides, I would assume you also have to change the shape? And the dim_names)
> 
> `ToTensor` currently amends `strides`, `shape` and `dim_names`.

Yes, but the `shape` and `dim_names` are defined as assuming row major layout, in which case the strides always need to reflect that? (or thus not passed, so Tensor::Make assumes that)

Reply:
I might have confused this, but I feel like if the outermost stride is manipulated inner strides remain constant. In our case we always just add the outermost stride. Maybe my mental model is off here, will verify.

Reply:
You are using `tensor_strides = this->strides();` (and then indeed prepend it with an additional stride for the extra length dimension). But so this are the _computed_ strides based on permutation, I think? (so they _are_ in some way manipulated) While the shape and dim_names are not permuted.

Reply:
> Yes, but the shape and dim_names are defined as assuming row major layout

If this is indeed the case then I think my original concern was unfounded. My initial interpretation of the spec was that `shape` and `dim_names` could be specified in an arbitrary (e.g. column-major) order, but `permutation` was what mapped those to the mandatory row-major physical layout. So, in that case, they would need to manually reordered before constructing the `Tensor` - since the tensor doesn't have any concept of permutations or logical vs physical layout.

It seems that it's the other way around? i.e. `shape` and `dim_names` must mirror the physical layout, but you can express a non-row-major logical layout with `permutation`. In that case, you shouldn't need to reorder anything before constructing the tensor. However, that would also mean that you _shouldn't_ permute the computed strides. I guess that would mean that `permutation` wouldn't affect any computations internal to `FixedShapeTensorType`. It would only be of descriptive use to the application.

>You are using `tensor_strides = this->strides();` (and then indeed prepend it with an additional stride for the extra length dimension). But so this are the computed strides based on permutation, I think? (so they are in some way manipulated) While the shape and dim_names are not permuted.

This is also true (but I didn't catch it). The strides are first computed using `shape`, but then they may be permuted. Therefore, the strides returned by `this->strides()` no longer correspond to the values in `shape`. (Of course, that would be solved by not permuting the strides, as described above).

Reply:
I've changed `ComputeStrides/this->strides()` to use `ComputeRowMajorStrides` on permuted `shape`. `ToTensor` now permutes `this->strides()` and prepends outermost stride (see commit https://github.com/apache/arrow/pull/8510/commits/14080dda44fc4b029a6cd83ed46a174ae2f953fd#). Also added a roundtrip test to check `ToTensor(FromTensor(tensor)) == tensor`.

Reply:
@jorisvandenbossche @benibus

Reply:
> I've changed `ComputeStrides/this->strides()` to use `ComputeRowMajorStrides` on permuted `shape`

Can you explain a bit what the consequence is of this? I see that tests for `strides()` changed its expected result, but the documentation of `strides()` didn't change?